teps/0138-decouple-api-and-feature-versioning.md

```diff
@@ -72,8 +91,331 @@ Per [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-
+
+### <a name="_tcstxie74non"></a>How many tests can we run in a reasonable amount of time?
+Based on a [recent PR](https://github.com/tektoncd/pipeline/pull/7032), our integration tests take between 26 mins (stable) → 33 mins (alpha). We don’t want to go beyond that. Based on [Feature flags testing matrix](https://docs.google.com/document/d/1r_MX9-mzRtdbfNQq5VC4guHb-tphA0WxWhlQIsusEEA/edit?resourcekey=0-RALry7-GaKn9i19UEaRnYg) benchmarking, the approximate time is:
+
+T = N<sub>pipelines</sub>\*N<sub>tasks</sub>\*N<sub>features</sub>\*6 s
+
+
+
+Assume: 
+
+N<sub>features</sub> = 20 (currently we have 17 api features; lets assume that at any point in time, we will likely have ~20 individual features)
+N<sub>tasks</sub> = 2 (two tasks per pipeline)
+
+
+|Scenario|N<sub>pipelines</sub>|N<sub>features</sub>|N<sub>tasks/pipeline</sub>|T (mins)|
+| :- | :- | :- | :- | :- |
+|How many pipelines can we afford to run in 30 mins?|**7**|20|2|30|
+|How long would it take to run a single (same) pipeline for all the features?|1|20|2|**4**|
+
+### How many tests should we run against the additional tests?
+- It is not feasible to run the entire e2e test suite against N features.
+- Coming up with a single test that covers all common core features might be challenging.
+  - We probably want our tests to be able to cover
+    - params
+    - workspaces
+    - results
+    - DAG (small fan out)?
+    - Finally
+    - remote resolution
+    - ...
+- We can afford to run a maximum of 7 pipelines per feature combination in a reasonable amount of time.
+- During implementation, we can figure out the number of pipelines that allow us to cover as many areas of the API spec as possible.
+
+For testing out individual per feature flags, we will use unit tests for each single feature flag in end to end tests and the combinations of features that could have overlapped(needs to elaborate). The integration tests for each flag will be in place for the CI while the combinations will be in the nightly tests.
+
+
+
+- Testing of behavioural features remains the way they run as of today.
+
+## Future Work
+- For seeking better way of communicating enabled features from cluster operators and downstream users including service providers and pipeline authors. CLI feature request to query all enabled features at each stability level via tkn.
+    - Example output:
+      ```
+      $pipeline: tkn features list enabled
+      stable: <feature-0>
+      beta: <feature-1>, <feature-2>
+      alpha: <feature-3>
+      ```
+- For preserving the ability of easily enabling all features of a stability level, we could provide some script that turns on all alpha/beta features by modifying the group of per feature flags.
+
+- In order to avoid feature To have the process of documenting the stability of features asn an auto genenrated process according to the Per Feature Flags struct being updated as a source of truth.
```

Comment:
```suggestion
- In order to avoid feature to have the process of documenting the stability of features an auto generated process according to the Per Feature Flags struct being updated as a source of truth.
```

I still do not follow this statement 🙃

Reply:
I think it means that the table of feature flags and their stability level should be generated automatically from the code. I think this is a good idea - a slightly different approach could be to use a json or yaml file to track the level of stability of the different features. The yaml could be used to generate docs as well as to select tests.

Reply:
Thanks for the question @pritidesai and clarification @afrittoli . Exactly, I wanted to say that this might be prone to error for `config-feature-flags` or just a Markdown files if people are editing it manually. Either a hack script or another json file could be very helpful in this case.

I also updated the wording for this future work.